NCSDK-32185: Add suit A/B sample documentation (#21213)
nrfconnect/sdk-nrf: samples/suit/ab/README.rst

.. _nrf54h_suit_ab_sample:

SUIT: Device firmware A/B update on the nRF54H20 SoC
####################################################

.. contents::
   :local:
   :depth: 2

The sample demonstrates how to configure the application, so it will be updated using the "A/B" method using :ref:`SUIT <ug_nrf54h20_suit_intro>` as well as the usage of manifest-controlled variables for implementing a mechanism to confirm device operation after the update.

The sample is capable of being updated using SMP protocol with both transports: UART or Bluetooth® Low Energy.

Requirements
************

The sample supports the following development kit:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf54h20dk_nrf54h20_cpuapp

You need the nRF Device Manager app for SUIT update over Bluetooth Low Energy:

* `nRF Device Manager mobile app for Android`_

>   (The minimum required version is v2.4.2.)

Author: I prefer to keep the same formatting as in the `smp_transfer` sample for consistency.

* `nRF Device Manager mobile app for iOS`_
  (The minimum required version is v1.9.3.)

> To perform a SUIT update over UART using the SMP protocol, install SUIT commands in nrfutil by running the ``nrfutil install suit`` command (requires nrfutil version 0.8.1 or higher).

Author: This part is  copied from the `smp_transfer` sample.

Memory layout definition
========================

The sample uses "A/B" method for updates, that has strong requirements on the partitioning of the system.
To make this part of the configuration easier, the whole memory map has been set using a single file, dedicated for the sample, located under ``sysbuild/nrf54h20dk_nrf54h20_memory_map.dtsi``.
It defines two memory regions for the main application: ``cpuapp_slot_b_partition`` and ``cpuapp_slot_a_partition`` as well as two memory regions for the radio stack: ``cpurad_slot_b_partition`` and ``cpurad_slot_a_partition``.

It is important to note a few rules, that were followed while writing the common memory map:

* The default MRAM partitioning has been completely removed by removing the nodes that group partitions, based on the assigned memory access permissions (``cpurad_rx_partitions``, ``cpuapp_rx_partitions`` and ``cpuapp_rw_partitions``).
* For each executable part of the sample, two equally sized partitions, one with ``a_partition``, and one with ``b_partition`` suffix were defined.

> * The partitions are still grouped in such a way, that partitions with the same access rights occupy a continuous region in the memory.

Author: I still find it important to mark that they are grouped.
Rephrased into:
```
* Partitions with the same access rights are grouped and occupy a continuous region in the memory.
```

  so enabling them would result in access requests from both domains to the same memory regions through their UICRs.
  The domain-specific overlays are going to include the common memory map and modify it by enabling the set of partitions, owned by the domain.
* A small staging area, called ``dfu_partition`` was defined.
  The size of it is set to 20 kB, so it is capable of holding all update candidate manifests, but is insufficient to store firmware binaries.
  This is acceptable, because firmware binaries will be written directly into their final destination.
* Both VPR cores (FLPR and PPR) are disabled in this sample. Adding VPR cores is out of scope of this documentation.

.. figure:: img/partitions.png
   :alt: Application and radio partitions on nRF54H20

   Application and radio partitions on nRF54H20

The common memory map has to be included in the image-specific overlay files.
This is done by adding the ``#include "nrf54h20dk_nrf54h20_memory_map.dtsi"`` inside:

* ``boards/nrf54h20dk_nrf54h20_cpuapp.overlay`` for the main application image
* ``sysbuild/hci_ipc.overlay`` for the radio stack image

> Apart from including the common memory map, a few additional lines, that enable the partitions that are owned by the respective domain has to be added:

Author: `the following lines` removed: I do not provide the code snippet here.

* ``-DCONFIG_<config_name>=<new_value>``: sets the ``<config_name>`` to the ``<new_value>`` in the primary variant (slot ``A``) of the ``<image>``.
  For example, the following parameter will change the number of LED blinks only in the variant ``A``:

   .. code-block:: console

      -DCONFIG_N_BLINKS=2

* ``-D<image>_CONFIG_<config_name>=<new_value>``: sets the ``<config_name>`` to the ``<new_value>`` in both variants of the ``<image>``.
  For example, the following parameter will change the number of LED blinks in both variants ``A`` and ``B``:

   .. code-block:: console

      -Dab_CONFIG_N_BLINKS=2

* ``-D<image>_slot_b_CONFIG_<config_name>=<new_value>``: sets the ``<config_name>`` to the ``<new_value>`` in the secondary variant (slot ``B``) of the ``<image>``.
  For example, the following parameter will change the number of LED blinks only in the variant ``B``:

   .. code-block:: console

      -Dab_slot_b_CONFIG_N_BLINKS=2

Manifest Provisioning Information
=================================

The second part of the sample configuration is the Manifest Provisioning Information (MPI).

> The default A/B sample configuration enables seven manifest classes:

Author: It was just another day, so I forgot and skipped it. Let's keep the quotes everywhere.

   -DSB_EXTRA_CONF_FILE="suit_mpi.conf"

Manifest templates
==================

The last part of the sample configuration are SUIT manifests.
For the "A/B" update purposes, the generic manifest templates are insufficient.
That's why the project includes the ``suit`` directory with the dedicated set of manifest templates.
For a better understanding of connection between manifest logic and manifest variables, it is worth to go through the logic, implemented in the root manifest sequences.

The logic inside the update sequences (suit-payload-fetch, suit-candidate-verification as well as suit-install) is sensitive to the set of available integrated (candidate) manifests.
If the Nordic top manifest is present, the logic will attempt to update the Nordic-controlled firmware.
If it is not available, the manifest will process application and radio manifests.

That said, it is impossible to update both Nordic firmware as well as application and radio firmware using a single update package.
Due to this limitation, all sequences participating in the update candidate processing validates the compatibility of the update package with the counterpart that is currently installed.

> Those conditions on versions can be configured using the ``VERSION`` file by setting:

Author: You're right - attributes fits better. Those are actually referenced by their name, effectively working as global variables.

* The value of ``SET_TO_UPDATE`` variable is copied to the ``SET_TO_UPDATE_MIRROR`` manifest variable.
  The ``SET_TO_UPDATE_MIRROR`` variable is stored in local SDFW RAM and cannot be updated by local cores.
* The suit-candidate-verification sequence from the local manifests is executed.
  The main purpose of them is to check that all of the staging areas (DFU partition, caches, as well as in-place updateable regions) were correctly populated with the data.
* If the Nordic manifest is present, the logic verifies Nordic firmware binaries stored inside the temporary DFU cache partition.

.. figure:: img/ab_root_candidate_verification.png
   :alt: The suit-candidate-verification sequence of the root manifest

   The suit-candidate-verification sequence of the root manifest

Once the suit-candidate-verification sequences are completed successfully, the SDFW continues the update procedure by executing the suit-install sequence of the root manifest.

The suit-install sequence
-------------------------

The logic inside the suit-install is very similar to the logic inside the suit-candidate-verification sequence.
The main differences are:

> * The suit-install sequence from the local manifests is executed.

Author: If you assume that it is obvious that the ``<x>`` sequence of the root manifest executes the ``<x>`` sequence in the local manifests, than yes, this point can be skipped.

It is now the applications responsibility to set this variable to ``CONFIRMED``.
Otherwise this variant will be considered as degraded on the next boot attempt.
The following activity diagram presents the logic, implemented inside the suit-invoke sequence.

.. figure:: img/ab_root_invoke.png
   :alt: The suit-invoke sequence of the root manifest

   The suit-invoke sequence of the root manifest

It is worth noticing that apart from regular application/radio/root manifest, the application recovery manifest is also defined.
In the A/B sample there is no separate recovery binary, but just the recovery manifest is used.
The recovery manifest is very similar to the root manifest, so in case of the root manifest damage, the recovery manifest serves as a backup logic.
The recovery manifest logic deviates from the root manifest in two areas:

* The boot preference variable is ignored and the manifest simply tries to boot first the variant ``A`` and then the variant ``B``.
* The recovery logic always boots the application and radio in the "degraded" mode.

User interface
**************

LED 0:

>     This LED is controlled by the application variant ``A``.

Reviewer: Just three empty spaces for the description texts.

Author: The same formatting mistake was made in the `smp_transfer` sample. Fixed only inside the a/b sample.

>     This LED is controlled as "active low", meaning it will turn on once the application is booted and blinks will be visible as short off periods.

Reviewer: ```suggestion
    This LED is controlled as "active low", meaning it will turn on once the application is booted and blinks in short intervals.
```

>     The ``CONFIG_N_BLINKS`` Kconfig option sets the number of blinks.

Reviewer: ```suggestion
    The :kconfig:option:`CONFIG_N_BLINKS` Kconfig option sets the number of blinks.
```

Reviewer: Is the ab_ suffix necessary for these sample-specific Kconfig options?

Author: If removed, causes conflicts with the Kconfigs defined in the `smp_transfer` sample.

>     The default of the Kconfig is set to ``1``, causing a single blink to indicate *Version 1*.

Reviewer: ```suggestion
    The default value of the option is set to ``1``, causing a single blink to indicate *Version 1*.
```

>     This value can be incremented to represent an update, such as setting it to ``2`` to indicate *Version 2*.

Reviewer: ```suggestion
    You can increment this value to represent an update, such as set it to ``2`` to indicate *Version 2*.
```

>     Using the ``-Dab_CONFIG_N_BLINKS`` command line option, the number of blinks can be set for both variants of the application.

Reviewer: ```suggestion
    Using the ``-Dab_CONFIG_N_BLINKS`` command-line option, you can set the number of blinks for both variants of the application.
```

    If the application variant ``A`` is started, this LED will not blink.

LED 2 and 3:
    These LEDs indicate the status of manifests inside the device:

    * If both LEDs are turned ``On``, the application is booted normally.
    * If both LEDs blink simultaneously, the application is booted in a degraded mode (i.e. into an unconfirmed variant).
    * If both LEDs blink in an alternating pattern, the application is booted using the recovery manifest.
    * If only LED 2 blinks and LED 3 is turned ``Off``, the application is booted without radio capabilities.

BUTTON 3:
    By pressing this button, the value of the boot preference setting, stored inside one of the manifest-controlled variables is changed.

Configuration
*************

|config|

The default configuration uses both UART and BLE with sequence number 1 (shown as Version 1 in the nRF Device Manager app).

To change the sequence number of the application, configure the ``APP_ROOT_SEQ_NUM`` inside the :file:`VERSION` file, used for :ref:`zephyr:app-version-details` in Zephyr and the |NCS|.

> It sets the :ref:`sequence number <ug_suit_dfu_suit_manifest_elements>` of the :ref:`SUIT envelope <ug_suit_dfu_suit_concepts>`’s manifest.

Author: FYI: copied from the `smp_transfer` sample docs.

   The default value is ``2``.

.. _ab_CONFIG_ID_VAR_CONFIRM_SET_B:

CONFIG_ID_VAR_CONFIRM_SET_B - Configuration for the Id of Manifest-controlled variable used by the root manifest, representing after-update confirmation state for image set B.
   The default value is ``3``.

.. _ab_CONFIG_EMULATE_APP_HEALTH_CHECK_FAILURE:

CONFIG_EMULATE_APP_HEALTH_CHECK_FAILURE - Configuration for blocking a confirmation of being healthy after the update.
   The default value is ``n``.

Building and running
********************

.. |sample path| replace:: :file:`samples/suit/smp_transfer`

This sample can be found under |sample path| in the |NCS| folder structure.

.. note::
   |54H_engb_2_8|

.. note::
    |sysbuild_autoenabled_ncs|

Building and programming using the command line
===============================================

To build and program the sample to the nRF54H20 DK, complete the following steps:

1. |open_terminal_window_with_environment|
#. Navigate to |sample path|.
#. Build the sample using the following command, with the following Kconfig options set:

   .. code-block:: console

>       west build -p -b nrf54h20dk/nrf54h20/cpuapp -- -DSB_EXTRA_CONF_FILE="suit_mpi.conf" -Dab_CONFIG_N_BLINKS=1

Author: If you do not use it - the sample will still work, it simply won't have all of the MPIs configured, as defined by the `suit_mpi.conf` file.
Since in the introductory we pointed out usage of this file as beneficial, I prefer to apply it in the testing steps.
User applications may not use `sample.yml` file, so at this stage I prefer to describe all options that should be used instead.

         sequenceNumber: 1
         semantic version: 0.1.0

   The application local 1 (``A`` variant) as well as the root manifests are now updated to the version ``2``.
   The **LED 0** should blink twice for every 5 seconds, indicating the second version of the application variant ``A``.

#. Revert to the previous firmware using the boot preference variable value.

   Press the BUTTON 3 on the DK.
   The application should switch the ``BOOT_PREFERENCE`` variable value:

   .. code-block:: console

      <inf> AB: Changing a boot preference (a -> B)
      <inf> AB: restart the device to enforce

#. Confirm that the image ``B`` still works.

   Reboot the DK using the reset button.
   After reboot the new root manifest will boot the old application variant ``B``.
   The following log message should appear on the logging serial interface:

   .. code-block:: console

      Boot preference: set B
      Boot status: image set B active
      Confirm status set A: confirmed
      Confirm status set B: confirmed

   The **LED 1** should blink once for every 5 seconds, indicating the first version of the application variant ``B``.

#. Enter the degraded mode.

>    It is possible to test the degraded mode using the sample.
>    If the application enables ``CONFIG_EMULATE_APP_HEALTH_CHECK_FAILURE`` then any update will not be confirmed, thus after two updates
>    the device will contain two ucncofirmed variants.

Reviewer: ```suggestion
   You can test the degraded mode using the sample.
   If the application enables the :ref:`ab_CONFIG_EMULATE_APP_HEALTH_CHECK_FAILURE <ab_CONFIG_EMULATE_APP_HEALTH_CHECK_FAILURE >` Kconfig option, no update is confirmed.
   Thus after two updates, the device contains two unconfirmed variants.
```

>    In such case, the logic inside the root manifest will set the ``BOOT_STATUS`` to ``BOOT_A_DEGRADED``.

Reviewer: ```suggestion
   In such case, the logic in the root manifest sets the ``BOOT_STATUS`` to ``BOOT_A_DEGRADED``.
```

>    To test it, first build the third version of the application, that does not confirm its correctness after any update:

Reviewer: ```suggestion
   To test it, first build the third version of the application.
   It does not confirm its correctness after any update:
```

>    #. Update the SUIT envelope sequence number, by changing the following line to the :file:`VERSION` file:

Reviewer: ```suggestion
   #. Update the SUIT envelope sequence number by changing the following line in the :file:`VERSION` file:
```

      .. code-block:: console

         APP_ROOT_SEQ_NUM = 3
         APP_LOCAL_1_SEQ_NUM = 3

>    #. Update the number of LED blinks, by rebuilding the sample with the following Kconfig options set:

Reviewer: ```suggestion
   #. Update the number of LED blinks by rebuilding the sample with the following Kconfig options set:
```

      .. code-block:: console

         west build -p -b nrf54h20dk/nrf54h20/cpuapp -- \
            -DSB_EXTRA_CONF_FILE="suit_mpi.conf" \
            -Dab_CONFIG_N_BLINKS=3 \
            -Dab_CONFIG_EMULATE_APP_HEALTH_CHECK_FAILURE=y

>       Another :file:`root.suit` file is created after running this command, that contains the updated firmware, that does not confirm its correctness.

Reviewer: ```suggestion
      Another :file:`root.suit` file is created after running this command.
      It contains the updated firmware, that does not confirm its correctness.
```

   Once created, send the update to the device twice:

   #. Change the directory to execute commands from DFU artifacts folder:

      .. code-block:: console

         cd build/DFU

>    #. Upload the image variant ``A`` with nrfutil:

Reviewer: ```suggestion
   #. Upload the image variant ``A`` using the following nrfutil command:
```

      .. code-block:: console

         nrfutil suit upload-envelope --serial-port COM7 --envelope-file root.suit --install --serve-images 60

      You should see an output similar to the following logged on UART::

      .. code-block::

         [00:01:01] ###### 100% [COM7] Uploaded
         ✔️ Serving images done
         ✔️ Uploaded hci_ipc.bin
         ✔️ Uploaded ab.bin

      The following log message should appear on the logging serial interface:

      .. code-block:: console

         <inf> AB: Image set A not confirmed yet, testing...
         <err> AB: App domain is NOT healthy
         <err> AB: Reboot the device to try to boot from previous firmware

>       The **LED 0** should blink three times for every 5 seconds, indicating the third version of the application variant ``A``.

Reviewer: ```suggestion
      The **LED 0** should blink three times every five seconds, indicating the third version of the application variant ``A``.
```

   #. Upload the image variant ``B`` with nrfutil:

      .. code-block:: console

         nrfutil suit upload-envelope --serial-port COM7 --envelope-file root.suit --install --serve-images 60

      You should see an output similar to the following logged on UART::

      .. code-block::

         [00:01:01] ###### 100% [COM7] Uploaded
         ✔️ Serving images done
         ✔️ Uploaded hci_ipc_variant_b.bin
         ✔️ Uploaded ab_variant_b.bin

      The following log message should appear on the logging serial interface:

      .. code-block:: console

         <inf> AB: Image set B not confirmed yet, testing...
         <err> AB: App domain is NOT healthy
         <err> AB: Reboot the device to try to boot from previous firmware

>       The **LED 1** should blink three times for every 5 seconds, indicating the third version of the application variant ``B``.

Reviewer: ```suggestion
      The **LED 1** should blink three times every five seconds, indicating the third version of the application variant ``B``.
```

   #. Reboot the DK using the reset button.

>       After reboot the root manifest will boot the unconfirmed application variant ``A`` in the degraded mode.

Reviewer: ```suggestion
      After the reboot, the root manifest boots the unconfirmed application variant ``A`` in the degraded mode.
```

      The following log message should appear on the logging serial interface:

      .. code-block:: console

         Boot preference: set B
         Boot status: image set A active, degraded mode
         Confirm status set A: not confirmed
         Confirm status set B: not confirmed

>       The **LED 0** should blink three times for every 5 seconds, indicating the third version of the application variant ``A``.

Reviewer: ```suggestion
      The **LED 0** should blink three times every five seconds, indicating the third version of the application variant ``A``.
```

      The **LED 2** and **LED 3** should blink simultaneously, indicating the degraded mode.

> #. Demage the radio firmware

Reviewer: ```suggestion
#. Degrade the radio firmware.
```
Did you mean degrade?

   Modify the variant ``A`` of the radio firmware to enter the degraded mode in the application variant ``B``.

   .. code-block:: console

      nrfutil device x-write --address 0xe054000 --value 0xFFFFFF --core Network --traits jlink

   You should see an output similar to the following logged on UART::

   .. code-block::

      ✔️ Data written to <snr>

   Reboot the DK using the reset button.

>    After reboot the root manifest will boot the unconfirmed application variant ``B`` in the degraded mode, with the radio.

Reviewer: ```suggestion
   After the reboot, the root manifest boots the unconfirmed application variant ``B`` in the degraded mode with the radio.
```

   The following log message should appear on the logging serial interface:

   .. code-block:: console

      Boot preference: set B
      Boot status: app image B active, degraded mode
      Confirm status set A: not confirmed
      Confirm status set B: not confirmed

   Modify the variant ``B`` of the radio firmware to enter the degraded mode without radio support.

   .. code-block:: console

>       nrfutil device x-write --address 0xe086000 --value 0xFFFFFF --core Network --traits jlink

Author: It should be fine.

   You should see an output similar to the following logged on UART::

   .. code-block::

      ✔️ Data written to <snr>

   Reboot the DK using the reset button.

>    After reboot the root manifest will boot the unconfirmed application variant ``A`` in the degraded mode, without the radio.

Reviewer: ```suggestion
   After the reboot, the root manifest boots the unconfirmed application variant ``A`` in the degraded mode without the radio.
```

   The following log message should appear on the logging serial interface:

   .. code-block:: console

      Boot preference: set B
      Boot status: app image A active, no radio, degraded mode
      Confirm status set A: not confirmed
      Confirm status set B: not confirmed

>    The **LED 0** should blink three times for every 5 seconds, indicating the third version of the application variant ``A``.
>    The **LED 2** should blink continuously.
>    The **LED 3** should be turned off.

Reviewer: ```suggestion
   * The **LED 0** should blink three times every five seconds, indicating the third version of the application variant ``A``.
   * The **LED 2** should blink continuously.
   * The **LED 3** should be switched off.
```

> #. Enter recovery mode by damaging root manifest.

Reviewer: ```suggestion
#. Enter the recovery mode by damaging the root manifest.
```
damaging?

Author: Changed to: "destroying"..

   Modify the root manifest to enter recovery mode.

   .. code-block:: console

>       nrfutil device x-write --address 0x0e1ef400 --value 0xFFFFFF --traits jlink

Author: This option is not supported:
```
nrfutil device x-write --address 0x0e1ef400 --value 0xFFFFFF --serial-port /dev/ttyACM0
error: unexpected argument '--serial-port' found

  tip: a similar argument exists: '--serial-number'

Usage: nrfutil device x-write --address <ADDRESS> --value <VALUE> <--serial-number <SERIAL_NUMBERS>|--traits <TRAITS>|--x-remote-jlink <REMOTE_JLINK_PROBES>|--x-append-batch <PATH>|--generate>

For more information, try '--help'.
```

   You should see an output similar to the following logged on UART::

   .. code-block::

      ✔️ Data written to <snr>

   Reboot the DK using the reset button.
   After reboot the root manifest will boot the unconfirmed application variant ``A`` in the degraded mode, without the radio, using recovery manifest.
   The following log message should appear on the logging serial interface:

   .. code-block:: console

      Boot in RECOVERY mode, probably the root manifest is damaged!
      Boot preference: set B
      Boot status: app image A active, no radio, degraded mode
      Confirm status set A: not confirmed
      Confirm status set B: not confirmed

   The **LED 0** should blink three times for every 5 seconds, indicating the third version of the application variant ``A``.

>    Both **LED 2** and **LED3** should blink in an alternating pattern, indicating the recovery mode.

Reviewer: ```suggestion
   Both **LED 2** and **LED 3** should blink in an alternating pattern, indicating the recovery mode.
```

#. Recover the device through the application update

>    To test it, first build the fourth version of the application, that confirm its correctness after any update:

Reviewer: ```suggestion
   To test it, first build the fourth version of the application, which confirms its correctness after an update:
```

   #. Update the SUIT envelope sequence number, by changing the following line to the :file:`VERSION` file:

      .. code-block:: console

         APP_ROOT_SEQ_NUM = 4
         APP_LOCAL_1_SEQ_NUM = 4

>    #. Update the number of LED blinks, by rebuilding the sample with the following Kconfig options set:

Reviewer: ```suggestion
   #. Update the number of LED blinks by rebuilding the sample with the following Kconfig options set:
```

      .. code-block:: console

         west build -p -b nrf54h20dk/nrf54h20/cpuapp -- \
            -DSB_EXTRA_CONF_FILE="suit_mpi.conf" \
            -Dab_CONFIG_N_BLINKS=4

>       Another :file:`root.suit` file is created after running this command, that contains the updated firmware, that confirms its correctness.

Reviewer: ```suggestion
      Another :file:`root.suit` file is created after running this command.
      It contains the updated firmware that confirms its correctness.
```

   #. Change the directory to execute commands from DFU artifacts folder:

      .. code-block:: console

         cd build/DFU

>    #. Upload the image variant ``B`` with nrfutil:

Reviewer: ```suggestion
   #. Upload the image variant ``B`` using the following nrfutil command:
```

      .. code-block:: console

         nrfutil suit upload-envelope --serial-port COM7 --envelope-file root.suit --install --serve-images 60

>       You should see an output similar to the following logged on UART::

Reviewer: ```suggestion
      You should see an output similar to the following logged on UART:
```

      .. code-block::

         [00:01:01] ###### 100% [COM7] Uploaded
         ✔️ Serving images done
         ✔️ Uploaded hci_ipc_variant_b.bin
         ✔️ Uploaded ab_variant_b.bin

      The following log message should appear on the logging serial interface:

      .. code-block:: console

         Boot preference: set B
         Boot status: image set B active
         Confirm status set A: not confirmed
         Confirm status set B: not confirmed

         <inf> AB: Image set B not confirmed yet, testing...
         <inf> AB: Confirming...
         <inf> AB: Confirmed

>       The **LED 1** should blink four times for every 5 seconds, indicating the fourth version of the application variant ``B``.
>       Both **LED 2** and **LED3** should be turned on, indicating that the device recovered from both recovery and degraded modes.

Reviewer: ```suggestion
      The **LED 1** should blink four times every five seconds, indicating the fourth version of the application variant ``B``.
      Both **LED 2** and **LED 3** should be switched on, indicating that the device recovered from both recovery and degraded modes.
```

   #. Upload the image variant ``A`` with nrfutil:

      .. code-block:: console

         nrfutil suit upload-envelope --serial-port COM7 --envelope-file root.suit --install --serve-images 60

>       You should see an output similar to the following logged on UART::

Reviewer: ```suggestion
      You should see an output similar to the following logged on UART:
```

      .. code-block::

         [00:01:01] ###### 100% [COM7] Uploaded
         ✔️ Serving images done
         ✔️ Uploaded hci_ipc.bin
         ✔️ Uploaded ab.bin

      The following log message should appear on the logging serial interface:

      .. code-block:: console

         Boot preference: set A
         Boot status: image set A active
         Confirm status set A: not confirmed
         Confirm status set B: confirmed

         <inf> AB: Image set A not confirmed yet, testing...
         <inf> AB: Confirming...
         <inf> AB: Confirmed

>       The **LED 0** should blink four times for every 5 seconds, indicating the fourth version of the application variant ``A``.
>       Both **LED 2** and **LED3** should be turned on.

Reviewer: ```suggestion
      The **LED 0** should blink four times every five seconds, indicating the fourth version of the application variant ``A``.
      Both **LED 2** and **LED 3** should be lit.
```

      .. code-block::

         manifest_info:
           classId: f03d385e-a731-5605-b15d-037f6da6097f (nRF54H20_nordic_top)
           semantic version: 0.9.5
           signatureCheck: signature check passed

Testing push mode with A/B sample
=================================

It is possible to configure the application to perform updates using the push method.

Since in the A/B method only a single variant is updateable for each boot (inactive one),
the PC tool must correctly handle the error codes, if the incorrect image is attempted to be transferred.

The images will be sent as DFU cache RAW images. Their IDs are defined using the common memory app, by adding the following aliases:

.. code-block:: dts

   dfu_target_img_22: cpuapp_slot_b_partition: partition@14c000 {
      reg = < 0x14c000 DT_SIZE_K(592) >;
   };

   dfu_target_img_21: cpuapp_slot_a_partition: partition@b8000 {
      reg = < 0xb8000 DT_SIZE_K(592) >;
   };

   ...

   dfu_target_img_32: cpurad_slot_b_partition: partition@86000 {
      reg = <0x86000 DT_SIZE_K(200)>;
   };

   dfu_target_img_31: cpurad_slot_a_partition: partition@54000 {
      reg = < 0x54000 DT_SIZE_K(200) >;
   };

> Using the declaration above:

Author: Rephrased to:
```
Using the declaration, the following images as RAW cache partitions are available:
```
